### PR TITLE
feat(http-global-proxy.js):  add http proxy feature to 4.2.x

### PR DESCRIPTION
**Pull request by kdhttps**

### CODE REVIEW

**Comment by kdhttps** [comment]
run tests

**Comment by christian-hawk** [comment]
run tests

**Comment by christian-hawk** [comment]
run tests

**Review by christian-hawk — APPROVED** [review]
@kdhttps thanks for that.
Please avoid a single commit for so many changes.
For every (single) test that pass, you do can do a commit.

File: server/utils/http-global-proxy.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const config = require('config')
+
+if (config.has('HTTP_PROXY')) {
+  const ga = require('global-agent')
+  ga.bootstrap()
+  global.GLOBAL_AGENT.HTTP_PROXY = config.get('HTTP_PROXY')
```
Comment (mookkiah) [inline server/utils/http-global-proxy.js:6]: Does this config read from operating system environment properties?

File: config/test.js [hunk 2]
```diff
@@ -184,6 +184,9 @@ const rateLimitWindowMs = 24 * 60 * 60 * 1000
 const rateLimitMaxRequestAllow = 100
 const cookieSameSite = 'none'
 const cookieSecure = true
+const HTTP_PROXY = 'http://localhost:3128'
+const HTTPS_PROXY = 'http://localhost:3129'
+const NO_PROXY = 'localhost,127.0.0.1'
```
Comment (mookkiah) [inline config/test.js:189]: Can we add a test case where NO_PROXY taking some regular expression like *.example.com

File: server/utils/http-global-proxy.js (hunk 1 above)
Comment (kdhttps) [reply server/utils/http-global-proxy.js:6]: right Mahendra, this config read from OS env properties. It supports every system architecture, just needs Environment properties in those systems.

File: config/test.js (hunk 2 above)
Comment (kdhttps) [reply config/test.js:189]: Sure, it supports regular expression

File: server/utils/http-global-proxy.js (hunk 1 above)
Comment (christian-hawk) [reply server/utils/http-global-proxy.js:6]: > Does this config read from operating system environment properties?
@mookkiah thanks for the contrib. This PR is already merged. Feel free to submit the PR or create this issue, your collaboration is much appreciated.

Comment (mookkiah) [reply server/utils/http-global-proxy.js:6]: Good. I think it expects different variable than standard ones...

https://github.com/gajus/global-agent#separate-proxy-for-https
https://github.com/gajus/global-agent#exclude-urls

@moabu 

Is it good idea to set this export transfer the value from standard ones like below.
```
HTTPS_PROXY=http://xxx.gluu.org:3128
HTTP_PROXY=http://xxx.gluu.org:3128
NO_PROXY=localhost,127.0.0.1,oxauth,oxtrust,oxshibboleth,opendj,jackrabbit,*.gluu.org,*.example.com
http_proxy=http://xxx.gluu.org:3128
https_proxy=http://xxx.gluu.org:3128
no_proxy=localhost,127.0.0.1,oxauth,oxtrust,oxshibboleth,opendj,jackrabbit,*.gluu.org,*.example.com
```

Maybe it is covered - https://github.com/gajus/global-agent#globalglobal_agent

Please test and confirm.

Comment (kdhttps) [reply server/utils/http-global-proxy.js:6]: > right Mahendra, this config read from OS env properties. It supports every system architecture, just needs Environment properties in those systems.

It is actually read from [here production.js](https://github.com/GluuFederation/gluu-passport/blob/558690123769a4ada6c5c313bd98fa85f6182221/config/production.js#L9).

OS Env --> production.js --> config.get('env')

Comment (kdhttps) [reply server/utils/http-global-proxy.js:6]: @mookkiah this feature supports standard properties, https://github.com/GluuFederation/gluu-passport/pull/340#issue-728744440



| Environment variable | Description |
|----------|-------------|
| HTTP_PROXY | Sets HTTP proxy to use |
| HTTPS_PROXY | Sets a distinct proxy to use for HTTPS requests.|
| NO_PROXY | Specifies a pattern of URLs that should be excluded from proxying.|

**Pull request merged**